### PR TITLE
feat(skore): Add data_source="both" to ComparisonReport metrics frames

### LINKED ISSUE #1874: enh(skore): Use train and test sets together via `data_source="both"` in Displays
### Is your feature request related to a problem? Please describe.

In skore v0.8, in the `plot_feature_importance` example ([link](https://docs.skore.probabl.ai/0.8/auto_examples/use_cases/plot_feature_importance.html)), for the `feature_importance` accessor of the `EstimatorReport`, we still have a lot of boilerplate code, such as plotting side by side train and test display to have easily the difference visible.  

### Describe the solution you'd like

The solution wouldn't be the same for all displays.  

For ROC and PR displays, we want to have the two curves on the same plot.  

For PredictionError, I'd say that we want all the points on the same plots, with different colors and/or points.  

For feature importance, we would have something as in the example:  

Image

### Describe alternatives you've considered, if relevant

We can have two subplots side by side.  
In particular for PredictionError, it can be a real solution.  

### Additional context

sub issue of https://github.com/probabl-ai/skore/issues/1416.

### LINKED ISSUE #2154: enh(skore): Implement option data_source="both" in PR curve Display for the ComparisonReport
### What would you like to say?

Part of https://github.com/probabl-ai/skore/issues/1874. 

As a reminder, today:
```python
# %%
from sklearn.datasets import load_breast_cancer
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from skore import train_test_split
from skore import EstimatorReport, ComparisonReport
X, y = load_breast_cancer(return_X_y=True)
split_data = train_test_split(X=X, y=y, random_state=0, as_dict=True)
classifier = LogisticRegression(max_iter=10_000)
report_1 = EstimatorReport(classifier, **split_data)
report_2 = EstimatorReport(RandomForestClassifier(), **split_data)
comp = ComparisonReport([report_1, report_2])
# %%
display = comp.metrics.precision_recall()
display.frame()
```

outputs:  

Image

and:  

Image

## Frame

For the frame, let's add an extra column data_source, and concatenate all the data in a long and unique dataframe.

## Plot

For the plot, to avoid having too many lines on the same plot (here in the example only two models are compared, but we could easily have more), and to avoid comparing the train curve of a model with the test curve of another, which wouldn't make any sense, let's use the subploting option once it's implemented in https://github.com/probabl-ai/skore/issues/1445 to have one plot with all the train curves, and one plot with all the test curves.  
Why not having one subplot per model with train and test at the same place? It's not necessarily a bad idea, and we could think about adding an option to offer both possibilities. Yet, to start simple and iterate, it seems that having one subplot per model is more useful when one wants to investigate one given model, and therefore it's the EstimatorReport job, rather than the ComparisonReport.

### LINKED ISSUE #2155: enh(skore): Implement option data_source="both" in ROC curve Display for the ComparisonReport
### What would you like to say?

Part of https://github.com/probabl-ai/skore/issues/1874. 

As a reminder, today:
```python
# %%
from sklearn.datasets import load_breast_cancer
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from skore import train_test_split
from skore import EstimatorReport, ComparisonReport
X, y = load_breast_cancer(return_X_y=True)
split_data = train_test_split(X=X, y=y, random_state=0, as_dict=True)
classifier = LogisticRegression(max_iter=10_000)
report_1 = EstimatorReport(classifier, **split_data)
report_2 = EstimatorReport(RandomForestClassifier(), **split_data)
comp = ComparisonReport([report_1, report_2])
# %%
display = comp.metrics.precision_recall()
display.frame()
```

outputs:  

Image

and:  

Image

## Frame

For the frame, let's add an extra column data_source, and concatenate all the data in a long and unique dataframe.

## Plot

For the plot, to avoid having too many lines on the same plot (here in the example only two models are compared, but we could easily have more), and to avoid comparing the train curve of a model with the test curve of another, which wouldn't make any sense, let's use the subploting option once it's implemented in https://github.com/probabl-ai/skore/issues/1445 to have one plot with all the train curves, and one plot with all the test curves.  
Why not having one subplot per model with train and test at the same place? It's not necessarily a bad idea, and we could think about adding an option to offer both possibilities. Yet, to start simple and iterate, it seems that having one subplot per model is more useful when one wants to investigate one given model, and therefore it's the EstimatorReport job, rather than the ComparisonReport.

### LINKED ISSUE #2156: enh(skore): Implement option data_source="both" in PredictionError Display for the ComparisonReport
### What would you like to say?

Part of https://github.com/probabl-ai/skore/issues/1874. 

## Frame

For the frame, let's add an extra column data_source, and concatenate all the data in a long and unique dataframe.

## Plot

For the plot, to avoid having too many scatter dots on the same plot (here in the example only two models are compared, but we could easily have more), and to avoid comparing the train curve of a model with the test curve of another, which wouldn't make any sense, let's use the subploting option once it's implemented in https://github.com/probabl-ai/skore/issues/1445 to have one plot with all the train curves, and one plot with all the test curves.  
Why not having one subplot per model with train and test at the same place? It's not necessarily a bad idea, and we could think about adding an option to offer both possibilities. Yet, to start simple and iterate, it seems that having one subplot per model is more useful when one wants to investigate one given model, and therefore it's the EstimatorReport job, rather than the ComparisonReport.

### PR DESCRIPTION
**Pull request by Adrien-1997**
Adds support for `data_source="both"` in ComparisonReport metric displays.

Train and test data are now stored in a single long DataFrame with an explicit
`data_source` column and plotted on separate subplots when requested, avoiding
ambiguous comparisons.

Applies to:
- Precision-Recall (Closes #2154)
- ROC (Closes #2155)
- Prediction Error (Closes #2156)

pe_train_test_plot
pr_train_test_plot
roc_train_test_plot

Closes #1874

### CODE REVIEW

**Comment by auguste-probabl** [comment]
Thanks for the PR! We are currently aiming to refactor this code using the seaborn API, which should make the logic quite a bit simpler (see #2193). We'll get back to this when we're done dealing with that!

**Comment by GaetandeCast** [comment]
Hey @Adrien-1997, we're coming to a close on the refactoring of the displays, with only [`roc_curve`](https://github.com/probabl-ai/skore/pull/2313) left to merge. I was wondering if you're still willing to go forward with this PR. If so you can start by resolving conflicts with main. If you don't have the availability, there is no issue, we can take over the PR. Just let us know!

**Comment by adrien-morel** [comment]
Thanks for the ping!  
Yes, I’m happy to move this PR forward. I’ll resolve the conflicts with `main` and push an updated version by Monday.

**Comment by GaetandeCast** [comment]
Thanks @Adrien-1997 ! Could you update the screenshots to show us how it looks in the `plot()`? Also could you add tests to make sure everything works as expected and avoid future regression ?

**Comment by adrien-morel** [comment]
Thanks Gaétan!

Yes, I’ll update the screenshots to reflect the new `plot()` output and add tests. I’ll push an update asap.

**Comment by adrien-morel** [comment]
Here’s what the plots look like with `data_source="both"`, and I’ve added unit tests.
pe_train_test_plot
pr_train_test_plot
roc_train_test_plot

File: skore/tests/unit/displays/precision_recall_curve/test_comparison_estimator.py [hunk 1]
```diff
@@ -501,3 +501,46 @@ def test_valid_subplot_by(fixture_name, subplot_by_tuples, request):
...
+    display.plot()
+
+    assert isinstance(display.ax_, (list, np.ndarray))
+    assert len(display.ax_) == len(report.reports_)
+
+    for ax in display.ax_:
+        assert isinstance(ax, mpl.axes.Axes)
+        assert len(ax.get_lines()) >= 1
+        assert ax.get_xlabel() == "recall"
+        assert ax.get_ylabel() in ("precision", "")
```
Comment (GaetandeCast) [inline skore/tests/unit/displays/precision_recall_curve/test_comparison_estimator.py]:
```suggestion
        legend = ax.get_legend()
        legend_texts = [text.get_text() for text in legend.get_texts()]
        assert len(legend_texts) == 2 
        assert "Train set" in legend_texts[0]
        assert "Test set" in legend_texts[1]
```

File: skore/tests/unit/displays/prediction_error/test_comparison_estimator.py [hunk 2]
```diff
@@ -322,3 +322,42 @@ def test_title(pyplot, linear_regression_with_train_test):
...
+    display.plot()
+
+    assert isinstance(display.ax_, (list, np.ndarray))
+    assert len(display.ax_) == len(report.reports_)
+
+    for ax in display.ax_:
+        assert isinstance(ax, mpl.axes.Axes)
+        assert len(ax.collections) >= 1
+        assert ax.get_xlabel() == "Predicted values"
+        assert ax.get_ylabel() == "Residuals (actual - predicted)"
```
Comment (GaetandeCast) [inline skore/tests/unit/displays/prediction_error/test_comparison_estimator.py]:
```suggestion
    legend_texts = [text.get_text() for text in display.figure_.legends[0].get_texts()]
    assert len(legend_texts) == 2 + 1 # 2 datasource + 1 perfect predictions
    assert legend_texts[0] == "train"
    assert legend_texts[1] == "test"
```

File: skore/tests/unit/displays/roc_curve/test_comparison_estimator.py [hunk 3]
```diff
@@ -482,3 +482,46 @@ def test_valid_subplot_by(fixture_name, subplot_by_tuples, request):
...
+    display.plot()
+
+    assert isinstance(display.ax_, (list, np.ndarray))
+    assert len(display.ax_) == len(report.reports_)
+
+    for ax in display.ax_:
+        assert isinstance(ax, mpl.axes.Axes)
+        assert len(ax.get_lines()) >= 1
+        assert ax.get_xlabel() == "False Positive Rate"
+        assert ax.get_ylabel() in ("True Positive Rate", "")
```
Comment (GaetandeCast) [inline skore/tests/unit/displays/roc_curve/test_comparison_estimator.py]:
```suggestion
        legend = ax.get_legend()
        legend_texts = [text.get_text() for text in legend.get_texts()]
        assert len(legend_texts) == 2 + 1 # 2 datasource + 1 chance level
        assert "Train set" in legend_texts[0]
        assert "Test set" in legend_texts[1]
```

File: skore/tests/unit/displays/prediction_error/test_comparison_estimator.py [hunk 4]
```diff
@@ -322,3 +322,42 @@ def test_title(pyplot, linear_regression_with_train_test):
     assert "Prediction Error" in title
     assert "estimator_1" not in title
     assert "Data source: Test set" in title
+
+
+def test_regression_data_source_both(pyplot, linear_regression_with_train_test):
+    """Regression test: `data_source='both'` should plot without crashing."""
+    estimator, X_train, X_test, y_train, y_test = linear_regression_with_train_test
```
Comment (auguste-probabl) [inline skore/tests/unit/displays/prediction_error/test_comparison_estimator.py]: Maybe a test fixture is in order here

**Comment by adrien-morel** [comment]
I’ve:
- Applied your review suggestions in both `prediction_error.py` and `_plot.utils.py` (same guard logic applied to `subplot_by="data_source"` to avoid invalid combinations in comparison + multiclass contexts),
- Accepted your proposed test changes, resolved the related conversations, and added a dedicated fixture as suggested by Auguste.

Thanks again for the detailed review and guidance 😊

File: skore/tests/unit/displays/prediction_error/test_comparison_estimator.py [hunk 5]
```diff
@@ -322,3 +322,47 @@ def test_title(pyplot, linear_regression_with_train_test):
     assert "Prediction Error" in title
     assert "estimator_1" not in title
     assert "Data source: Test set" in title
+
+
+@pytest.fixture
+def linear_regression_comparison_report(linear_regression_with_train_test):
+    """Fixture providing a ComparisonReport with two linear regression estimators."""
+    estimator, X_train, X_test, y_train, y_test = linear_regression_with_train_test
```
Comment (GaetandeCast) [inline skore/tests/unit/displays/prediction_error/test_comparison_estimator.py]: Let's move this to `conftest.py` where the other fixtures are defined.

**Review by GaetandeCast — CHANGES_REQUESTED** [review]
Last change then it's looking good for me!

**Comment by adrien-morel** [comment]
I’ve moved the comparison regression fixture to `conftest.py` with the other shared fixtures and updated the tests accordingly.

Also wanted to say I really enjoyed this review process, thanks a lot @GaetandeCast !

**Review by GaetandeCast — APPROVED** [review]
You're welcome! Let's see if @glemaitre wants to add something here before we merge.
